packages/xod-project/src/flatten.js

```diff
@@ -626,6 +626,143 @@ const removeTerminalsAndPassPins = ([nodes, links]) => {
+        )
+      )
+    )(nodes);
+
+    if (terminalNodeIds.length === 0) return [nodes, links];
+
+    const linksConnectedToTerminals = R.filter(
+      R.compose(
+        linkNodeIds => R.any(R.contains(R.__, linkNodeIds), terminalNodeIds),
+        Link.getLinkNodeIds
+      ),
+      links
+    );
+
+    const lowermostTerminalLinks = R.filter(
+      R.compose(
+        R.complement(R.contains)(R.__, terminalNodeIds),
+        Link.getLinkInputNodeId
+      ),
+      linksConnectedToTerminals
+    );
+
+    // For each lowermost link find the full chain to the uppermost link and
+    // colleact all links and NodeIds in the arrays.
+    // Links collected in order from lowermost to uppermost, so first and last
+    // links in array will be used to create a new link.
+    // NodeIds collected to omit Nodes that replaced with a new link.
+    // :: [[NodeId, Link]]
+    const terminalChains = R.map(lowermostLink =>
+      traverseUpAndCollectTerminalChain(
+        terminalNodeIds,
+        linksConnectedToTerminals,
+        [[], [lowermostLink]]
+      )
+    )(lowermostTerminalLinks);
+
+    const directLinks = R.compose(
+      R.reject(R.isNil),
+      R.map(([_, _links]) => {
+        if (_links.length <= 1) return null;
+
+        const lowermost = R.head(_links);
+        const uppermost = R.last(_links);
+
+        const lowermostId = Link.getLinkId(lowermost);
+        const uppermostId = Link.getLinkId(uppermost);
+
+        const inputPinKey = Link.getLinkInputPinKey(lowermost);
+        const inputNodeId = Link.getLinkInputNodeId(lowermost);
+        const outputPinKey = Link.getLinkOutputPinKey(uppermost);
+        const outputNodeId = Link.getLinkOutputNodeId(uppermost);
+
+        return R.compose(
+          R.assoc('id', `${uppermostId}-to-${lowermostId}`),
+          Link.createLink
+        )(inputPinKey, inputNodeId, outputPinKey, outputNodeId);
+      })
+    )(terminalChains);
+
+    const nodeIdsToOmit = R.compose(R.flatten, R.pluck(0))(terminalChains);
+    const linkIdsToOmit = R.compose(
+      R.map(Link.getLinkId),
+      R.flatten,
+      R.pluck(1)
+    )(terminalChains);
+
+    const nodesWithoutTerminals = R.reject(
+      R.compose(R.contains(R.__, nodeIdsToOmit), Node.getNodeId),
+      nodes
+    );
+    const linksWithoutTerminalLinks = R.reject(
+      R.compose(R.contains(R.__, linkIdsToOmit), Link.getLinkId),
+      links
+    );
+
+    const newNodesAndLinks = R.map(
+      splitLinkWithCastNode(project, patchTuples, nodesWithoutTerminals),
+      directLinks
+    );
+
+    return R.compose(
+      R.over(
+        R.lensIndex(0),
+        R.compose(R.concat(nodesWithoutTerminals), R.reject(R.isNil))
+      ),
+      R.over(R.lensIndex(1), R.concat(linksWithoutTerminalLinks)),
+      // Convert array of pairs `[ [Node, [Link]] ]` into pair of flat arrays `[ [Node], [Link] ]`
+      R.values,
+      R.applySpec([R.pluck(0), R.compose(R.flatten, R.pluck(1))])
+    )(newNodesAndLinks);
```

Comment:
I'm fairly sure it can be replaced with
```
return R.converge(R.pair, [
  R.compose(R.concat(nodesWithoutTerminals), R.reject(R.isNil), R.pluck(0)),
  R.compose(R.concat(linksWithoutTerminalLinks), R.flatten, R.pluck(1)),
])(newNodesAndLinks)
```

similar to like it's done in `createCastNodes `